# Workflow of using VCF-kit to design primers related to polymorphic restriction site in vcf files
Nan Hu / 
Nov. 22, 2020

---

## Installation and environment settings for VCF-kit (important)
To prepare for installing VCF-kit, make sure system has pip and anaconda installed ahead. 
If new users do not have these two tools installed, you can refer to [pip installing guide](https://pip.pypa.io/en/stable/installing/) 
and [anaconda installing guide](https://docs.anaconda.com/anaconda/install/linux/) to get instructions for installation. Once required tools are prepared, 
execute below commands in commandline:
```bash
conda config --add channels bioconda
conda config --add channels conda-forge
conda create -n vcf-kit \
  danielecook::vcf-kit=0.2.6 \
  "bwa>=0.7.17" \
  "samtools>=1.10" \
  "bcftools>=1.10" \
  "blast>=2.2.31" \
  "muscle>=3.8.31" \
  "primer3>=2.5.0"
```
These process a virtual enviroment only for VCF-kit. We can activate this enviroment by:
```bash
conda activate vcf-kit
```
Since this software have not been updated for several years, there are some python codes not accommodating new biopython package. 
We need to install old versions of biopython and fix some issues there.
```bash
# uninstall new version of biopython
conda uninstall biopython
# install v1.77 biopython
conda install biopython=1.77
# uninstall numpy thoroughtly, repeat this step until there is nothing to uninstall
conda uninstall numpy
# re-install numpy
conda install numpy
```
Check the availability of VCF-kit with our expected function by:
```bash
vk primer --help
```
Last step to prepare the environment is to fix some coding issue in finding primer3 software. 
VCF-kit expected we installed Primer3 under system folder so by default it will search primer3-config files in some of these folders.
We need to find where our primer3-config files are and manually add to the searching range.
```bash
find ~ -iname primer3_config -type d
```
There will be a result at ```~/anaconda3/lib/python3.6/site-packages/vcfkit/static/primer3_config```. Save this path and do another search:
```bash
find ~ -iname primer3.py -type f
```
*Noticed here, it is possible that you cannot find this config file due to new primer3 version. If so, you can use path of mine: ```/home/nan/anaconda3/lib/python3.6/site-packages/vcfkit/static/primer3_config```. I did not see any errors by using others config file right now.*

Use text editors (like: nano, vi etc.) to open the file locate at ```~/anaconda3/envs/vcf-kit/lib/python3.7/site-packages/vcfkit/utils/primer3.py```.
In this python script, find 'class primer3'. Under \# Global default, there is a block of script:
```python
       # Global default
        thermo_paths = ["/usr/local/share/primer3_config/",
                        "/usr/local/share/primer3/primer3_config/",
                        "~/.linuxbrew/share/primer3_config/",
                        "~/.linuxbrew/share/primer3/primer3_config/",
                        "/.linuxbrew/share/primer3_config/",
                        "/.linuxbrew/share/primer3/primer3_config/",
                        primer3_config]
```
Add our saved path to this constant default. Completed one will look like:
```python
        # Global default
        thermo_paths = ["/usr/local/share/primer3_config/",
                        "/usr/local/share/primer3/primer3_config/",
                        "~/.linuxbrew/share/primer3_config/",
                        "~/.linuxbrew/share/primer3/primer3_config/",
                        "/.linuxbrew/share/primer3_config/",
                        "/.linuxbrew/share/primer3/primer3_config/",
                        "~/anaconda3/lib/python3.6/site-packages/vcfkit/static/primer3_config",
                        primer3_config]
```
Now, we have finished enviroment settings for VCF-kit primer function.

## VCF preparation
VCF is a standard file type for variants calling containing basical information of polymorphic sites in population.
It can be generated from multiple softwares. In our lab, we use GATK 4.0 to call genotypes and filter the vcf file.
For further information of .vcf file type please refer to [Variant Call Format](https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format).

VCF-kit reads contig/chromosome length by reading the tag of sequence ID in vcf files. However, vcf files generated by GATK having the reference genome file name after
the sequence length, which will be incorrectly read read as length variants so that causing an error. Before we run the command, those tags should be fixed.
Assuming our vcf file name is ```SE.parent.vcf```.
```bash
sed 's/,assembly=Salix_purpurea_var_94006.mainGenome.fasta//g' SE.parent.vcf > SE.parent.fixedname.vcf
```
Since we use Salix purpurea as our reference genome, the replacement regular expression is like above. 
If using different reference genomes or different file names, it has better to check the ID tags before by ```grep "##contig=" SE.parent.vcf``` and replace exact everything after 'length=<Number>' until the last '>'.

When finishing replacing, we need to compress the file into .gz format for VCF-kit as an input. There is no need to index the new vcf file since the software will do it if there is no index.
```bash
bgzip -c SE.parent.fixedname.vcf > SE.parent.fixedname.vcf.gz
```

##







